snakemake/shell.py

```diff
@@ -79,6 +103,20 @@ def cleanup(cls):
         with cls._lock:
             cls._processes.clear()
 
+    @classmethod
+    def use_bash_on_win(cls, bashcmd=None):
```

Comment:
I am not yet convinced that it is a good idea to expose this via the public API. Why not just installing bash via conda for the windows tests?

Reply:
I agree that this shouldn't really be in the public API. It is mostly something we need to easily make the tests run on windows. I.e. so snakemake runs all the shell arguments with bash instead of cmd.exe which is default. 

 We could make the method private and then only use it in snakemakes internal tests. An other option I could think of is an argument to snakemake it self and the test system could launch test with that argument set so all the test were run with bash as the shell instead of cmd.exe.  That would also remove the `shell.use_bash_on_win()` from all the test `snakefile`s

Reply:
> Why not just installing bash via conda for the windows tests?

We still need a way to tell snakemake to bash instead of cmd.exe to interpret shell commands. 

I think I like the idea with a argument to snakemake.

either:
```
snakemake --use-bash-on-windows
```
or a general command line argument which could work for both windows/posix
```
snakemake --shell-executable=<path-to-bash.exe>
```
The test framework could then just set this argument when running test on windows.

Reply:
Brainstorming:

We add a directive to the language:
```
shell: "bash"
```
We could restrict possible values to "bash" and "cmd.exe"?

Modify the linter to advise to always add this to a workflow.

And if that is set, [this](https://github.com/snakemake/snakemake/blob/master/snakemake/shell.py#L176) code path will be also taken on windows, thereby failing with a proper error if bash is not in PATH.

Reply:
@johanneskoester, I like the idea. It is much cleaner. 

The only downside is that "bash.exe" is on path by default in Windows. 

![image](https://user-images.githubusercontent.com/1038978/77823762-c73e3100-70fd-11ea-83c3-4efbe940e6c3.png)

But it is a version that can't be used. It is just a link for starting the windows subsystem for linux. It will just give some weird errors for all users, until they realize that the need to set the PATH manually. But maybe that is OK. The main point of all this work is to get as much of the test suite running on windows. 

Could we also allow `shell:` to take a full path? So you do something like this:

```
shell: "path/to/a/good/bash.exe"
```
or 
```
shell: locate_a_good_bash()
```

How do I add a new shell directive to the language? Or could you do that, and then I could rebase and then use the directive?

Reply:
@johanneskoester I you could add the "shell:" directive in this branch then I will look at getting the whole thing merged with master again.

Reply:
Again thinking about this. Snakemake already offers:
```
shell.executable("bash")
```
Which could be done in all the tests, or even in the run function of tests/common.py. Hence, on windows you can simply put your bash finding code into tests/common.py, and set `shell.executable` to whatever full path is determined there. Does that work? This way, we don't expose the method in the regular Snakemake API. 
So, either put something like 
```
if ON_WINDOWS:
    shell.executable(find_bash_on_windows())
```
into the run function of tests/common.py. And also move the function `find_bash_on_windows()` there. This way, we don't need to touch all the tests.

Reply:
Yes. I can see how that would work for tests that are started directly by pytest. 

But setting stuff in `common.py` would have no effect on the sub work flows i.e. the inner work flows where snakemake starts it self.

Reply:
Maybe we could get pytest to modify the Windows `PATH` so the good version of `bash` is found first?

Reply:
Yes, that's a good idea, can you do that please?

Reply:
I will. But give me some days. I also need to fix a lot of test regressions.